src/main/cpp/src/cast_string_to_datetime.cu

```diff
@@ -857,6 +860,169 @@ std::unique_ptr<cudf::column> parse_ts_strings(cudf::strings_column_view const&
     num_rows, std::move(output_columns), /* null_count */ 0, rmm::device_buffer(), stream, mr);
 }
 
+/**
+ * Parse date string to year, month, day.
+ * Note: Spark date supports max 7 digits year, Spark timestamp supports max 6 digits year.
+ */
+__device__ bool parse_date(unsigned char const* const ptr,
+                           unsigned char const* const ptr_end,
+                           spark_rapids_jni::date_segments& d)
+{
+  int pos     = 0;
+  int end_pos = ptr_end - ptr;
+
+  // trim left
+  while (pos < end_pos && is_whitespace(ptr[pos])) {
+    ++pos;
+  }
+
+  // trim right
+  while (pos < end_pos && is_whitespace(ptr[end_pos - 1])) {
+    --end_pos;
+  }
+
+  if (eof(pos, end_pos)) { return false; }
+
+  // parse sign
+  bool negative_year_sign = false;
+  char const sign_c       = ptr[pos];
+  if ('-' == sign_c || '+' == sign_c) {
+    ++pos;
+    if ('-' == sign_c) { negative_year_sign = true; }
+  }
+
+  // parse year: yyyy[y][y][y]
+  if (!parse_int(ptr,
+                 pos,
+                 end_pos,
+                 d.year,
+                 /*min_digits*/ 4,
+                 /*max_digits*/ 7)) {
+    return false;
+  }
+  if (negative_year_sign) { d.year = -d.year; }
+
+  if (eof(pos, end_pos)) {
+    // only has: yyyy[y][y][y], return early
+    return true;
+  }
+
+  // parse month: -[m]m
+  if (!parse_char(ptr, pos, '-') || !parse_int(ptr,
+                                               pos,
+                                               end_pos,
+                                               d.month,
+                                               /*min_digits*/ 1,
+                                               /*max_digits*/ 2)) {
+    return false;
+  }
+
+  if (eof(pos, end_pos)) {
+    // only has: yyyy[y][y]-[m]m, return early
+    return true;
+  }
+
+  // parse day: -[d]d
+  if (!parse_char(ptr, pos, '-') || !parse_int(ptr,
+                                               pos,
+                                               end_pos,
+                                               d.day,
+                                               /*min_digits*/ 1,
+                                               /*max_digits*/ 2)) {
+    return false;
+  }
+
+  if (eof(pos, end_pos)) {
+    // no tailing characters, e.g.: "2025-01-01"
+    return true;
+  }
+
+  // parse date time separator
+  // ignore the tailing chars after ' ' or 'T', e.g.: the following are valid:
+  // "2025-01-01T", "2025-01-01Txxxx", and "2025-01-01 xxx"
+  return parse_date_time_separator(ptr, pos);
+}
+
+struct parse_string_to_date_fn {
+  // input strings
+  cudf::column_device_view d_strings;
+
+  // output columns
+  uint8_t* null_mask;
+  cudf::timestamp_D* output;
+
+  __device__ void operator()(cudf::size_type const idx) const
+  {
+    auto const str                   = d_strings.element<cudf::string_view>(idx);
```

Comment:
No need for null check?

Reply:
Yes, not need. But I updated, checking null will be a little faster IMO.

```cpp
    // check null
    if (d_strings.is_null(idx)) {
      validity[idx] = false;
      return;
    }
```